### PR TITLE
MI-220 File caching for user profile picture

### PR DESCRIPTION
**Pull request by limphilip**
A file caching is required for files that can be long to convert from the response of the server, such as the facebook user profile picture. Two options of caching were available:
- **Memory Caching**: Allows for faster retrieval of cached items, but less storage capacity.
- **Disk Storage Caching**: Allows to cache larger amount of items, but has slower performance compared to memory caching.

Memory caching was chosen, because the main goal here is to boost the performance of the caching. Although the storage capacity is limited (to 4MB in this case, but can easily be updated), the capacity can still hold about 26 user different profile pictures, which seems reasonable for the mRides application.

**Here is a summary of how the caching works**:
The caching mimics the Android LRU cache. The cache is represented by a dictionary in which the key is the facebook ID of the user associated with the picture being cached, and the value is a bitmap object representing the user profile picture. 

Due to the limited size of the cache, we need to decide which element is to be removed from the cache whenever it is full, and a new element needs to be added. The cache will remove the element that has not been accessed for the longest period of time. This is handled by a list of facebook user ID stored in which the beginning of the list represents the elements that were most recently accessed, and the end contains the elements that were accessed a long time ago.

**Integrating the cache with existing components**:
The `UserMapper` is in charge of retrieving the profile picture, either from the server or from the cache. Thus, the `UserMapper` first looks in the cache if the picture of a user is present, if it is, it will use it. Only when it is absent from the cache, that the mapper will ask the gateway to obtain it from the web.

The `UserMapper` asks the `UserCache` for the picture. Since the `UserCache` is already responsible for retrieving other types of cached information such the preferences of the user, it will have visibility over the new `MemoryCache` package, which is the actual implementation of the mimicked LRU cache.

**Testing**:
Unit tests and integration tests have been written in order to ensure that the cache functions properly. Those tests tests that functionalities of adding, removing and retrieving elements from the cache. They also ensure that the list of most recently accessed works properly, and that correct elements are removed from the cache when the cache is full and a new element needs to be added.

Also related to #47

### CODE REVIEW

File: UnitTests/CacheTest/MemoryCacheTest/UserFacebookProfilePictureCacheTest.cs [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+            cache.Add(facebookId, imageBitmap);
+
+            // Also add it to the list of recently accessed
+            List<long> listRecentlyAccessed = GetObjectField<List<long>>(userFbPicCache, "listUserFacebookIdRecentlyAccessed");
+            listRecentlyAccessed.Add(facebookId);
+
+            // Check what we can find from the cache
+            Bitmap bitmapFound = userFbPicCache.FindUserFacebookProfilePicture(facebookId);
+            Assert.True(bitmapFound != null);
+            Assert.True(bitmapFound.Equals(imageBitmap));
```
Comment (AlineTheAlien) [inline UnitTests/CacheTest/MemoryCacheTest/UserFacebookProfilePictureCacheTest.cs]: It's bad practice to have a test case with more than one "assert", you might want to combine these into one "assert" statement. It helps pinpoint why a test case failed, and it follows the good quality assurance standards 👍 
Besides this the test cases look really good!

Comment (limphilip) [reply UnitTests/CacheTest/MemoryCacheTest/UserFacebookProfilePictureCacheTest.cs]: Thanks @AlineTheAlien , I will update this.

**Comment by limphilip** [comment]
@AlineTheAlien I have updated the test cases and also updated myself with dev. Let me know if you can find any other issues, otherwise I think it's good to merge. Thanks again for pointing out the test cases 😃

**Comment by AlineTheAlien** [comment]
@limphilip There seems to be a merge conflict in the app config file preventing me from merging, should I undo your last push (the merge of dev into your branch)? I think this branch was already safe to merge because there were no conflicts with dev so I think it's fine to merge this pull request without merging dev into it first.

**Comment by limphilip** [comment]
It seems like my local dev wasn't up to date with the one on the server. I will update this branch again, and push when it's ready.

**Comment by limphilip** [comment]
@AlineTheAlien It is up to date now, the conflict was due to files being added to the csproj, so they got added to the same lines. I took both the one from dev and mine so that all files are present when loading the visual studio project.

**Comment by AlineTheAlien** [comment]
Ok looks good! Merging now 👍

**Pull request merged**